cub/cub/warp/specializations/warp_exchange_shfl.cuh

```diff
@@ -0,0 +1,327 @@
+      //         ┌──────┐        ┌──────┐
+      //  A4  A5 │A6  A7│ E4  E5 │E6  E7│
+      //         │      │        │      │
+      //         │      │        │      │
+      //  B4  B5 │B6  B7│ F4  F5 │F6  F7│
+      //         └──────┘        └──────┘
+      // ┌──────┐        ┌──────┐
+      // │C4  C5│ C6  C7 │G4  G5│ G6  G7
+      // │      │        │      │
+      // │      │        │      │
+      // │D4  D5│ D6  D7 │H4  H5│ H6  H7
+      // └──────┘        └──────┘
+      //
+      //       ROUND 3
+      //       =======
+      //  t0  t1  t2  t3  t4  t5  t6  t7
+      //     ┌──┐    ┌──┐    ┌──┐    ┌──┐
+      //  A0 │A1│ C0 │C1│ E0 │E1│ G0 │G1│    NUM_ENTRIES == 1 so we have 4
+      //     └──┘    └──┘    └──┘    └──┘       submatrices per thread and there
+      // ┌──┐    ┌──┐    ┌──┐    ┌──┐           is 1 element between these
+      // │B0│ B1 │D0│ D1 │F0│ F1 │H0│ H1        submatrices.
+      // └──┘    └──┘    └──┘    └──┘
+      //     ┌──┐    ┌──┐    ┌──┐    ┌──┐  1. t[0,2,4,6] data[1] swap with t[1,3,5,7]'s data[0]
+      //  A2 │A3│ C2 │C3│ E2 │E3│ G2 │G3│  2. t[0,2,4,6] data[3] swap with t[1,3,5,7]'s data[2]
+      //     └──┘    └──┘    └──┘    └──┘  3. t[0,2,4,6] data[5] swap with t[1,3,5,7]'s data[4]
+      // ┌──┐    ┌──┐    ┌──┐    ┌──┐      4. t[0,2,4,6] data[7] swap with t[1,3,5,7]'s data[6]
+      // │B2│ B3 │D2│ D3 │F2│ F3 │H2│ H3
+      // └──┘    └──┘    └──┘    └──┘
+      //     ┌──┐    ┌──┐    ┌──┐    ┌──┐
+      //  A4 │A5│ C4 │C5│ E4 │E5│ G4 │G5│
+      //     └──┘    └──┘    └──┘    └──┘
+      // ┌──┐    ┌──┐    ┌──┐    ┌──┐
+      // │B4│ B5 │D4│ D5 │F4│ F5 │H4│ H5
+      // └──┘    └──┘    └──┘    └──┘
+      //     ┌──┐    ┌──┐    ┌──┐    ┌──┐
+      //  A6 │A7│ C6 │C7│ E6 │E7│ G6 │G7│
+      //     └──┘    └──┘    └──┘    └──┘
+      // ┌──┐    ┌──┐    ┌──┐    ┌──┐
+      // │B6│ B7 │D6│ D7 │F6│ F7 │H6│ H7
+      // └──┘    └──┘    └──┘    └──┘
+      //
+      //       RESULT
+      //       ======
+      //  t0  t1  t2  t3  t4  t5  t6  t7
+      //
+      //  A0  B0  C0  D0  E0  F0  G0  H0
+      //
+      //
+      //  A1  B1  C1  D1  E1  F1  G1  H1
+      //
+      //
+      //  A2  B2  C2  D2  E2  F2  G2  H2
+      //
+      //
+      //  A3  B3  C3  D3  E3  F3  G3  H3
+      //
+      //
+      //  A4  B4  C4  D4  E4  F4  G4  H4
+      //
+      //
+      //  A5  B5  C5  D5  E5  F5  G5  H5
+      //
+      //
+      //  A6  B6  C6  D6  E6  F6  G6  H6
+      //
+      //
+      //  A7  B7  C7  D7  E7  F7  G7  H7
+      //
+
+      // NOTE: Do *NOT* try to refactor this code to use a reference, since nvcc
+      //       tends to choke on it and then drop everything into local memory.
+      const InputT send_val = (xor_bit_set ? CompileTimeArrayImpl<OutputT, IDX>::val
+                                           : CompileTimeArrayImpl<OutputT, IDX + NUM_ENTRIES>::val);
+      const InputT recv_val = __shfl_xor_sync(mask, send_val, NUM_ENTRIES, ITEMS_PER_THREAD);
```

Comment:
I know that it's not important at the moment, but still, since the `width` is in terms of threads, do you think it's worth rewriting the line as below?
```suggestion
      const InputT recv_val = __shfl_xor_sync(mask, send_val, NUM_ENTRIES, LOGICAL_WARP_THREADS);
```

Reply:
done